### PR TITLE
Fix [`multiple_crate_versions`] to correctly normalize package names to avoid missing the local one

### LINKED ISSUE #12145: `multiple_crate_versions` doesn't report when name contains a dash
### Summary

When using `multiple_crate_versions` on a crate with an dash in the name the `multiple_crate_versions` does not trip.

If you use `multiple_crate_versions` on a crate without a `-`, so either a `_` or just no `-` at all, the lint works as expected.

Is it because  `LOCAL_CRATE` ([link](https://github.com/rust-lang/rust-clippy/blob/a71211d0b52c01f1b37fe544e0e13fd1bdc31979/clippy_lints/src/cargo/multiple_crate_versions.rs#L13)) is normalized, i.e. `-` is replaced by `_`. Further down, comparing it to `p.name` ([link](https://github.com/rust-lang/rust-clippy/blob/a71211d0b52c01f1b37fe544e0e13fd1bdc31979/clippy_lints/src/cargo/multiple_crate_versions.rs#L19)) that is an issue as `p.name` has the package name (with `-` intact), not the normalized name.

Fix: https://github.com/rust-lang/rust-clippy/pull/12146

### Reproducer

Setup 1 (with dash in name):
```bash
cargo new with-dash
cd with-dash
cargo add regex stubr --quiet
cargo tree --depth 1
cargo clippy --quiet -- --deny clippy::multiple_crate_versions
```

Output 1:
```bash
kristof in ~/__SOURCES/temp
❯ cargo new with-dash
     Created binary (application) `with-dash` package

kristof in ~/__SOURCES/temp
❯ cd with-dash

kristof in with-dash on  main [?] is 📦 v0.1.0 via 🦀 v1.75.0
❯ cargo add regex stubr --quiet

kristof in with-dash on  main [?] is 📦 v0.1.0 via 🦀 v1.75.0
❯ cargo tree --depth 1
with-dash v0.1.0 (/home/kristof/__SOURCES/temp/with-dash)
├── regex v1.10.2
└── stubr v0.6.2
    [build-dependencies]
```

Setup 2 (with underscore in name):

```bash
cargo new with_underscore
cd with_underscore
cargo add regex stubr --quiet
cargo tree --depth 1
cargo clippy --quiet -- --deny clippy::multiple_crate_versions
```

Output 2:
```bash
kristof in ~/__SOURCES/temp
❯ cargo new with_underscore
     Created binary (application) `with_underscore` package

kristof in ~/__SOURCES/temp
❯ cd with_underscore

kristof in with_underscore on  main [?] is 📦 v0.1.0 via 🦀 v1.75.0
❯ cargo add regex stubr --quiet

kristof in with_underscore on  main [?] is 📦 v0.1.0 via 🦀 v1.75.0
❯ cargo tree --depth 1
with_underscore v0.1.0 (/home/kristof/__SOURCES/temp/with_underscore)
├── regex v1.10.2
└── stubr v0.6.2
    [build-dependencies]

kristof in with_underscore on  main [?] is 📦 v0.1.0 via 🦀 v1.75.0
❯ cargo clippy --quiet -- --deny clippy::multiple_crate_versions
error: multiple versions for dependency `async-channel`: 1.9.0, 2.1.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions
  = note: requested on the command line with `-D clippy::multiple-crate-versions`

error: multiple versions for dependency `async-io`: 1.13.0, 2.2.2
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `async-lock`: 2.8.0, 3.3.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `base64`: 0.13.1, 0.21.7
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `bitflags`: 1.3.2, 2.4.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `event-listener`: 2.5.3, 4.0.3
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `fastrand`: 1.9.0, 2.0.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `futures-lite`: 1.13.0, 2.2.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `getrandom`: 0.1.16, 0.2.12
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `linux-raw-sys`: 0.3.8, 0.4.12
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `polling`: 2.8.0, 3.3.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `rand`: 0.7.3, 0.8.5
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `rand_chacha`: 0.2.2, 0.3.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `rand_core`: 0.5.1, 0.6.4
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `regex-syntax`: 0.6.29, 0.8.2
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `rustix`: 0.37.27, 0.38.30
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `socket2`: 0.4.10, 0.5.5
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `syn`: 1.0.109, 2.0.48
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `wasi`: 0.9.0+wasi-snapshot-preview1, 0.11.0+wasi-snapshot-preview1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows-sys`: 0.48.0, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows-targets`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_aarch64_gnullvm`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_aarch64_msvc`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_i686_gnu`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_i686_msvc`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_x86_64_gnu`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_x86_64_gnullvm`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_x86_64_msvc`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: could not compile `with_underscore` (bin "with_underscore") due to 28 previous errors
```

Setup 3 (with nothing in the name, just one word):

```bash
cargo new justoneword
cd justoneword
cargo add regex stubr --quiet
cargo tree --depth 1
cargo clippy --quiet -- --deny clippy::multiple_crate_versions
```

Output 3:
```bash
kristof in ~/__SOURCES/temp
❯ cargo new justoneword
     Created binary (application) `justoneword` package

kristof in ~/__SOURCES/temp
❯ cd justoneword

kristof in justoneword on  main [?] is 📦 v0.1.0 via 🦀 v1.75.0
❯ cargo add regex stubr --quiet

kristof in justoneword on  main [?] is 📦 v0.1.0 via 🦀 v1.75.0
❯ cargo tree --depth 1
justoneword v0.1.0 (/home/kristof/__SOURCES/temp/justoneword)
├── regex v1.10.2
└── stubr v0.6.2
    [build-dependencies]

kristof in justoneword on  main [?] is 📦 v0.1.0 via 🦀 v1.75.0
❯ cargo clippy --quiet -- --deny clippy::multiple_crate_versions
error: multiple versions for dependency `async-channel`: 1.9.0, 2.1.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions
  = note: requested on the command line with `-D clippy::multiple-crate-versions`

error: multiple versions for dependency `async-io`: 1.13.0, 2.2.2
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `async-lock`: 2.8.0, 3.3.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `base64`: 0.13.1, 0.21.7
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `bitflags`: 1.3.2, 2.4.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `event-listener`: 2.5.3, 4.0.3
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `fastrand`: 1.9.0, 2.0.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `futures-lite`: 1.13.0, 2.2.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `getrandom`: 0.1.16, 0.2.12
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `linux-raw-sys`: 0.3.8, 0.4.12
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `polling`: 2.8.0, 3.3.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `rand`: 0.7.3, 0.8.5
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `rand_chacha`: 0.2.2, 0.3.1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `rand_core`: 0.5.1, 0.6.4
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `regex-syntax`: 0.6.29, 0.8.2
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `rustix`: 0.37.27, 0.38.30
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `socket2`: 0.4.10, 0.5.5
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `syn`: 1.0.109, 2.0.48
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `wasi`: 0.9.0+wasi-snapshot-preview1, 0.11.0+wasi-snapshot-preview1
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows-sys`: 0.48.0, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows-targets`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_aarch64_gnullvm`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_aarch64_msvc`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_i686_gnu`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_i686_msvc`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_x86_64_gnu`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_x86_64_gnullvm`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: multiple versions for dependency `windows_x86_64_msvc`: 0.48.5, 0.52.0
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_crate_versions

error: could not compile `justoneword` (bin "justoneword") due to 28 previous errors
```

### Version

```text
❯ rustc -Vv
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: x86_64-unknown-linux-gnu
release: 1.75.0
LLVM version: 17.0.6
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by kristof-mattei**
Fixes #12145 

changelog: [`multiple_crate_versions`]: correctly normalize package name

### CODE REVIEW

File: clippy_lints/src/cargo/multiple_crate_versions.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ pub(super) fn check(cx: &LateContext<'_>, metadata: &Metadata) {
 
     if let Some(resolve) = &metadata.resolve
         && let Some(local_id) = packages.iter().find_map(|p| {
-            if p.name == local_name.as_str() {
+            if p.name.replace('-', "_") == local_name.as_str() {
```
Comment (Manishearth) [inline clippy_lints/src/cargo/multiple_crate_versions.rs:19]: I wonder if we can avoid the allocation by using iteration here

I wonder if that's necessary though

Comment (kristof-mattei) [reply clippy_lints/src/cargo/multiple_crate_versions.rs]: Reading more: crates.io only allows a subset of ASCII:
https://github.com/rust-lang/crates.io/blob/825cce6c189fcf1174da4b7d31855bae430ed97c/src/models/keyword.rs#L56-L63

Cargo itself only allows ASCII:
https://github.com/rust-lang/cargo/blob/92395d90106b3b61bcb68bcf2069052c93771764/src/cargo/ops/cargo_new.rs#L260C32-L266

But I still believe `char` is better.

Comment (Manishearth) [reply clippy_lints/src/cargo/multiple_crate_versions.rs]: > While this allocates a char on each end per char compared at least it stops once they're not equal.

it's a stack allocation, not usually what people mean when talking about allocations.

Comment (Manishearth) [reply clippy_lints/src/cargo/multiple_crate_versions.rs]: but yeah why not make the change to use `.as_bytes()` and normalize `b'-'` to `b'_'`

Comment (kristof-mattei) [reply clippy_lints/src/cargo/multiple_crate_versions.rs]: > but yeah why not make the change to use `.as_bytes()` and normalize `b'-'` to `b'_'`

Let me do that.

Comment (kristof-mattei) [reply clippy_lints/src/cargo/multiple_crate_versions.rs]: @Manishearth Done. Thanks for the mentoring.

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
@bors r+

thanks!

**Pull request merged**